bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/KarafAddonService.java

```diff
@@ -146,9 +147,18 @@ private Addon getAddon(Feature feature) {
                 break;
         }
 
+        // for openHAB add-on bundles the package is the same as the bundle name
+        List<String> packages = feature.getBundles().stream().filter(bundle -> !bundle.isDependency()).map(bundle -> {
```

Comment:
Do we really need a list of packages for an add-on? We usually have a strict naming convention that all classes of an add-on should reside under the same (parent) package. The only exception is usually 3rd-party code that is packaged within or as `src/3rdparty/java`, but as this would be in the same bundle, this would not be covered by the logic here.

Thinking about how this feature can be supported in the UI, I'd think it would be best, if there's a single toggle through which the user sets the log level for the add-on. Wdyt?

Reply:
Yes, the user should only see a single switch. I'm not good at UI design, but my idea would be that in the add-ons overview (Settings/Bindings or similar) a drop-down element is displayed next to the "remove" button (and not shown when the button is labelled "installed" as it makes no sense to change loggers for an uninstalled add-on). The drop-down should contain the log-levels (with a default "DEFAULT" and TRACE/DEBUG/INFO/WARN/ERROR as other options). So the packages itself are never shown to the user.

It's true that add-ons that are provided by the `KarafAddonService` cannot make use of a list at the moment. Other implementations of `AddonService` (like the `JsonAddonService`) may have different logic to determine the packages (and also support add-ons which do not follow the strict naming policy that openHAB uses). IMO the `List` is more future-proof and does not hurt if it is only a single bundle.

Reply:
My concern is that the UI would be forced to have more complex logic, as it can't simply do a single REST call when logging level is changed for an add-on, but it would have to read the package list and do multiple REST calls sequentially. That's imho not nice to move such logic to the UI, it should be handles in the core, if nothing is expected to be seen by the user anyway. It might be more elegant to have a `setLogLevel` method on the `AddonService` itself - which could then be independently implemented by the different implementations and which could at the same time be used by e.g. console commands or rule actions, which might be future additions. Wdyt?

Reply:
That would be elegant from the API point of view (and in fact my first idea), but there are several reasons why I decided against than:

- We would need a dependency to a karaf bundle in each `AddonService` because it is needed to set the log level. Currently we have decoupled karaf from that.
- We would need to duplicate the logging code (three times).
- It would be API breaking because we would need to require each service to implement the new method.

It would be possible to workaround the first point by introducing a new OSGi service `LogLevelService` as an optional dependency. That‘s not very nice either.

I would suggest another option: we could make the REST API accept either a list of logger names and a single level (the UI can then just re-use the result from the first call) or a list of logger names/log levels (more flexible but requires some processing on client side). I would go with the first one.

Reply:
I did check the code again, and the UI can already change all loggers in one request: The list of logger names/log-levels is already implemented. So all the UI would have to do is process

```
["some.nice.package", "another.nice.package"]
```

to

```
[
  { "loggerName": "some.nice.package", "level" : "DEBUG" },
  { "loggerName": "another.nice.package", "level" : "DEBUG" }
]
```

which I believe is a one-liner in most languages (e.g. JS: `let loggerSettings = loggerNames.map(v => ({"loggerName" : v, "level" : "DEBUG"}));`).

Reply:
Ok, thanks for your explanations, let's go for it then and hope that @ghys will make use of it in the UI. 😄

Reply:
I would try doing it myself, unfortunately I can't build the UI repo due to https://github.com/openhab/openhab-webui/issues/1165.